### PR TITLE
[Uid] Unable to extend Uuid/Ulid and use fromString()

### PR DESCRIPTION
**Pull request by OskarStark**
| Q             | A
| ------------- | ---
| Branch?       | 5.1
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | ---
| License       | MIT
| Doc PR        | -

Before:
```php
CustomUuid::fromString(...) // returns `Uuid`
```

After:
```php
CustomUuid::fromString(...) // returns `CustomUuid`
```

same for `Ulid` class.

cc @fancyweb

### CODE REVIEW

File: src/Symfony/Component/Uid/Tests/Fixture/CustomUlid.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Uid\Tests\Fixture;
+
+use Symfony\Component\Uid\Ulid;
+
+final class CustomUlid extends Ulid
+{
+    public function __construct(string $ulid = null)
```
Comment (fancyweb) [inline src/Symfony/Component/Uid/Tests/Fixture/CustomUlid.php]: Why is this needed?

File: src/Symfony/Component/Uid/Tests/Fixture/CustomUlid.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Uid\Tests\Fixture;
```
Comment (fancyweb) [inline src/Symfony/Component/Uid/Tests/Fixture/CustomUlid.php]: In most components, we named the dir "Fixtures".

**Comment by OskarStark** [comment]
Travis failure unrelated

File: src/Symfony/Component/Uid/Uuid.php [hunk 3]
```diff
@@ -74,7 +74,7 @@ public static function fromString(string $uuid): parent
             case NilUuid::TYPE: return new NilUuid();
         }
 
-        return new self($uuid);
+        return new static($uuid);
```
Comment (fancyweb) [inline src/Symfony/Component/Uid/Uuid.php]: This place can only be reached if the actual class is Uuid so it is safe here, there is nothing to change.

File: src/Symfony/Component/Uid/Ulid.php [hunk 4]
```diff
@@ -79,7 +79,7 @@ public static function fromString(string $ulid): parent
             base_convert(substr($ulid, 27, 5), 16, 32)
         );
 
-        return new self(strtr($ulid, 'abcdefghijklmnopqrstuv', 'ABCDEFGHJKMNPQRSTVWXYZ'));
+        return new static(strtr($ulid, 'abcdefghijklmnopqrstuv', 'ABCDEFGHJKMNPQRSTVWXYZ'));
```
Comment (fancyweb) [inline src/Symfony/Component/Uid/Ulid.php]: This place is only reached if the input is binary, the test needs to reflect that.

**Review by fancyweb — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Uid/Uuid.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Uid/Uuid.php]: this is already done L59 above

Comment (OskarStark) [reply src/Symfony/Component/Uid/Uuid.php]: reverted

File: src/Symfony/Component/Uid/Tests/Fixture/CustomUlid.php (hunk 2 above)
Comment (OskarStark) [reply src/Symfony/Component/Uid/Tests/Fixture/CustomUlid.php]: Thanks, renamed

File: src/Symfony/Component/Uid/Uuid.php (hunk 3 above)
Comment (OskarStark) [reply src/Symfony/Component/Uid/Uuid.php]: Reverted

File: src/Symfony/Component/Uid/Tests/Fixture/CustomUuid.php [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Uid\Tests\Fixture;
+
+use Symfony\Component\Uid\Uuid;
+
+final class CustomUuid extends Uuid
+{
+    public function __construct(string $uuid)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Uid/Tests/Fixture/CustomUuid.php]: no need for any constructor

File: src/Symfony/Component/Uid/Tests/Fixture/CustomUlid.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Uid/Tests/Fixture/CustomUlid.php]: not needed

File: src/Symfony/Component/Uid/Tests/Fixture/CustomUuid.php (hunk 5 above)
Comment (OskarStark) [reply src/Symfony/Component/Uid/Tests/Fixture/CustomUuid.php]: reverted

File: src/Symfony/Component/Uid/Tests/Fixture/CustomUlid.php (hunk 1 above)
Comment (OskarStark) [reply src/Symfony/Component/Uid/Tests/Fixture/CustomUlid.php]: reverted

**Review by fancyweb — APPROVED** [review]

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Uid/Ulid.php (hunk 4 above)
Comment (OskarStark) [reply src/Symfony/Component/Uid/Ulid.php]: Done

**Comment by nicolas-grekas** [comment]
Thank you @OskarStark.
